lib/Conversion/TorchToTcp/Elementwise.cpp

```diff
@@ -105,6 +105,72 @@ class ConvertAtenMulOp : public OpConversionPattern<AtenMulTensorOp> {
   }
 };
 
+class ConvertAtenBatchNormOp : public OpConversionPattern<AtenBatchNormOp> {
+public:
+  using OpConversionPattern<AtenBatchNormOp>::OpConversionPattern;
+  using OpAdaptor = typename AtenBatchNormOp::Adaptor;
+
+  LogicalResult
+  matchAndRewrite(AtenBatchNormOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Value input = adaptor.getInput();
+    RankedTensorType inputType = input.getType().dyn_cast<RankedTensorType>();
+
+    Value weight = adaptor.getWeight();
+    RankedTensorType weightType = weight.getType().dyn_cast<RankedTensorType>();
+
+    Value bias = adaptor.getBias();
+    RankedTensorType biasType = bias.getType().dyn_cast<RankedTensorType>();
+
+    Value running_mean = adaptor.getRunningMean();
+    RankedTensorType runningMeanType = running_mean.getType().dyn_cast<RankedTensorType>();
+
+    Value running_var = adaptor.getRunningVar();
+    RankedTensorType runningVarType = running_var.getType().dyn_cast<RankedTensorType>();
+
+    RankedTensorType resultType =
+        OpConversionPattern<AtenBatchNormOp>::getTypeConverter()
+            ->convertType(op.getType())
+            .template cast<RankedTensorType>();
+
+    if (!inputType || !weightType || !biasType ||
+          !runningMeanType || !runningVarType || !resultType)
+      return rewriter.notifyMatchFailure(
+          op, "Only Ranked Tensor types are supported in TCP");
+
+    assert(runningMeanType.getNumElements() != 0 && runningVarType.getNumElements() != 0);
```

Comment:
Is this check in the verifier for `AtenBatchNormOp`?

Reply:
No AFAICT. `AtenBatchNormOp` [uses](https://sourcegraph.com/github.com/llvm/torch-mlir/-/blob/include/torch-mlir/Dialect/Torch/IR/GeneratedTorchOps.td?L4556-L4557) `AnyTorchOptionalTensorType` for running mean/variance. Depending on the mode (training/inference) the op is used in and whether it has been exposed to any training cycles at all, it is possible these are never initialized. At the moment we only support inference mode batch_norm lowering and when running stats are available, hence this check.

From PyTorch [docs](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html) for BatchNorm2d:
```
track_running_stats ([bool](https://docs.python.org/3/library/functions.html#bool)) – a boolean value that when set to True, this module tracks the running mean and variance, and when set to False, this module does not track such statistics, and initializes statistics buffers running_mean and running_var as None. When these buffers are None, this module always uses batch statistics, in both training and eval modes. Default: True
```

Reply:
In that case, we should probably error out gracefully saying that is unsupported, instead of asserting here.

I think that is what @sanjoy was referring to as well - we can assert here if this was a check in the verifier of `AtenBatchNormOp`. If not, we should return an error instead.

Reply:
Ah makes sense - will switch it to a `rewriter.notifyMatchFailure` here.

Reply:
Done